lib/traject/psulib_marc.rb

```diff
@@ -56,3 +58,99 @@ def process_genre(record, fields)
   end
   genres
 end
+
+ESTIMATE_TOLERANCE = 15
+MIN_YEAR = 500
+MAX_YEAR = Time.new.year + 6
+
+# Feed the record to the MarcPubDateProcessor class to find the publication year.
+#
+# Refactor of Traject::Macros::Marc21Semantics#marc_publication_date as the basic logic but check for 264|*1|c before
+# 260c.
+def process_publication_date(record)
+  return nil unless record.is_a? MARC::Record
+
+  field008 = Traject::MarcExtractor.cached('008').extract(record).first
+  marc_date_processor = MarcPubDateProcessor.new(field008)
+  pub_date = marc_date_processor.find_date
+
+  if pub_date.nil?
+    # Nothing from 008, try 264 and 260
+    field264c = Traject::MarcExtractor.cached('264|*1|c', separator: nil).extract(record).first
+    field260c = Traject::MarcExtractor.cached('260c', separator: nil).extract(record).first
+    pub_date = marc_date_processor.check_elsewhere field264c, field260c
+  end
+
+  # Ignore dates below min_year (default 500) or above max_year (this year plus 6 years)
+  pub_date && (pub_date > MIN_YEAR || pub_date < MAX_YEAR) ? pub_date : nil
+end
+
+# For determining the single three to four digit publication year.
+#
+# If 008 represents a date range, will take the midpoint of the range, only if range is smaller than estimate_tolerance,
+# default 15 years.
+#
+# @param field008 [String] The 008 field from the Marc record
+# @return nil if field008 is not set
+# @return [Integer] if the field's date can be found, the year found
+class MarcPubDateProcessor
+  attr_accessor :field008, :date_type, :date1, :date2
+
+  def initialize(field008)
+    @field008 = field008
+    return unless @field008 && @field008.length >= 11
+
+    @date_type = field008.slice(6)
+    @date1_str = field008.slice(7, 4)
+    @date2_str = field008.length > 15 ? field008.slice(11, 4) : ''
+  end
+
+  # Based on the date_type, return the proper value.
+  def find_date
+    return nil if @date_type.nil? || @date_type == 'n'
+
+    case @date_type
+    when 'p', 'r'
+      # Reissue/reprint/re-recording, etc.
+      date_str = @date2_str.to_i != 0 ? @date2_str : @date1_str
+      resolve_date date_str
+    when 'q'
+      # Questionable
+      resolve_range
+    else
+      # Default case, just resolve the first date.
+      resolve_date @date1_str
+    end
+  end
+
+  # For when we are dealing with ranges.
+  def resolve_range
+    # Make unknown digits at the beginning or end of range
+    date1 = @date1_str.sub('u', '0').to_i
+    date2 = @date2_str.sub('u', '9').to_i
+    # Do we have a range we can use?
+    is_range = (date2 > date1) && ((date2 - date1) <= ESTIMATE_TOLERANCE)
+    (date2 + date1) / 2 if is_range
```

Comment:
@cdmo I added two more specs for `date_type=q`, one is failing. I want to see what you think about that case.

Reply:
🤔 

```
(byebug) @date1_str
"19uu"
(byebug) @date1_str.sub('u', '0')
"190u"
```

Reply:
also, it's returning `'2002'` because the `@record` has a `260c`. Is that intentional? If it has a 260c, shouldn't we expect it to return that?

Reply:
Yes, I agree it cannot find an acceptable date in 008, then it should return that 260c. But marc_publication_date is returning nil interestingly 🤔 I think it does not look correct that way, right?

Reply:
`marc_publication_date`? That's gone, did you mean `marc_date_processor.find_date`? That should return nil. But then the next step finds the 260c and returns that value, which is 2002, and the test is expecting nil.

Reply:
I was referring to  Traject's marc_publication_date which is using `Traject::Macros::Marc21Semantics.publication_date`. And yeah we are not using that anymore but I just tested against that and wanted to compare results.

Reply:
Yup yup I gotchya now, Slack convo cleared this up.

Reply:
Per our conversation on slack, I will change this spec to test with a fixture that does not have a 264 neither 260. And other than that I think our logic for this case is working correctly - meaning it resorts to 264 or 260 if there is one when there is no date found in 008 due to ESTIMATE_TOLERANCE check failing.